[torch] Add cuda support for segment reduction 'max'

serhaty opened the pull request:
Summary:
Building on top of previous PR. This PR adds cuda support for 1D max reduction.

Next steps:
- Add support for other major reduction types (e.g. min, sum) for 1D tensor
- Documentation for the op
- Perf optimizations and benchmark util
- Backward support  (not high priority)
- Support for multi dimensional tensors (on data and lengths) (not high priority)
- Support for 'indices' (not high priority)

Test Plan: Added unit test

Differential Revision: D27121170

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -0,0 +1,124 @@
+        auto temp_storage = at::empty(
+            {static_cast<int64_t>(temp_storage_bytes)},
+            data.options().dtype(kByte));
+        error = cub::DeviceSegmentedReduce::Reduce(
```
Please wrap all cub calls in AT_CUDA_CHECK instead of later internal assert (TORCH_INTERNAL_ASSERT means "it's a bug in pytorch", whereas here an error can mean one of many different things wrong with the gpu state):
```
AT_CUDA_CHECK(cub::DeviceSegmentedReduce::Reduce...
);
```

ngimel commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9119,4 +9119,4 @@
   variants: function
   dispatch:
-    CPU: _segment_reduce_cpu
+    CPU, CUDA: _segment_reduce_
```
trailing underscore usually means inplace operation, so please don't use it

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -80,5 +60,49 @@ Tensor _segment_reduce_cpu(
+
+  // length related checks
+  TORCH_CHECK(
+      lengths.has_value() && !indices.has_value(),
```
Do you have a follow-up to support indices-based reduction? That would allow you to not compute cumsum

serhaty replied:
next steps are roughly in order are:
- support more reduction types
- add backward support
- add more documentation/testing/benchmark
- add support for multi dimension input
- add support for indices based reduction

So it might take a while to come back to indices since not very high priority.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -80,5 +60,49 @@ Tensor _segment_reduce_cpu(
+
+Tensor _segment_reduce_(
+    const Tensor& data,
+    std::string reduce,
```
const

serhaty replied:
I tried to fix this but then the binding was not working properly, it complains with "undefined symbol"

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -0,0 +1,124 @@
+    const c10::optional<Tensor>& lengths,
+    int64_t axis,
+    bool unsafe) {
+  const auto& lengths_value = lengths.value();
```
you cannot have unchecked `.value()` access, you need to either check `has_value`, or have non-optional argument

serhaty replied:
I was checking on caller side already but yeah, no need to pass optional here actually.

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -0,0 +1,124 @@
+  const auto& lengths_value = lengths.value();
+  if (!unsafe) {
+    TORCH_CHECK(
+        (lengths_value >= 0).all().item<int64_t>(),
```
length_value.min()>0 is faster (we've agreed that there shouldn't be 0-lengths segments, right?)

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -0,0 +1,124 @@
+  template <typename OutputT>
+  __host__ __device__ __forceinline__ OutputT
+  operator()(const OutputT& a, const OutputT& b) {
+    return at::_isnan(b) ? b : std::max<OutputT>(a, b);
```
unfortunately, on ROCm there's a bug that doesn't correctly return first value from std::max, so here you have to check both a and b for _isnan

serhaty replied:
Thanks for the heads up!

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -0,0 +1,124 @@
+      at::cuda::getCurrentCUDAStream());
+  TORCH_INTERNAL_ASSERT(error == cudaSuccess, "Failed to make cub call!");
+
+  auto offset_temp_storage = at::empty(
```
It's a bit faster to do 
```
auto& allocator = *::c10::cuda::CUDACachingAllocator::get();
auto dataPtr = allocator.allocate(temp_storage_bytes);
```
and use `dataPtr.get()` in the call

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -0,0 +1,124 @@
+      "segment_reduce_cuda",
+      [&]() {
+        auto offsets = _get_complete_sum(lengths_contig);
+        auto* offsets_data_ptr = offsets.data_ptr<int64_t>();
```
2 lines above can be outside of dispatch macro (they don't need scalar_t)

ngimel reviewed (commented):
This looks good, I left a few comments

ezyang commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -80,5 +60,49 @@ Tensor _segment_reduce_cpu(
+    &_segment_reduce_cpu_kernel);
+REGISTER_AVX_DISPATCH(_segment_reduce_stub, &_segment_reduce_cpu_kernel);
+REGISTER_AVX2_DISPATCH(_segment_reduce_stub, &_segment_reduce_cpu_kernel);
+REGISTER_VSX_DISPATCH(_segment_reduce_stub, &_segment_reduce_cpu_kernel);
```
What the heck is going on here? If you don't put the CPU kernel in the cpu/ kernel you won't actually compile it for the different AVX/AVX2 variants. Was this copy pasted from somewhere? What are you trying to do?

ngimel replied:
You won't compile it for AVX/AVX2 variant, but dispatch stub would still be looking for those, so you have to register non-avx kernel for all of them, it won't compile otherwise.

ezyang replied:
Following up from @ngimel and I sync: This seems to be the current practice for some kernels, but to reduce boilerplate, structured kernels is better, see https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

ngimel replied:
Looks like you also need `REGISTER_ARCH_DISPATCH` here, for ibm power.

ngimel reviewed (approved):
lgtm

Commit:
[torch] Add cuda support for segment reduction 'max' (#54175)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/54175

Building on top of previous PR. This PR adds cuda support for 1D max reduction.

Next steps:
- Add support for other major reduction types (e.g. min, sum) for 1D tensor
- Documentation for the op
- Perf optimizations and benchmark util
- Backward support  (not high priority)
- Support for multi dimensional tensors (on data and lengths) (not high priority)
- Support for 'indices' (not high priority)

Test Plan: Added unit test

Reviewed By: ngimel

Differential Revision: D27121170

fbshipit-source-id: d8ea79905396ef8d09d11b49a75ac066bc2caa78
